qiime/sdk/execution.py

```diff
@@ -0,0 +1,44 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2016--, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+
+import os
+import tempfile
+import subprocess
+
+
+class ExecutionContext:
+
+    def __call__(self):
+        # Should this be decorated with abc.abstract_method?
+        pass
+
+
+class LocalSubprocessExecution:
+
+    _python_executable = "python"
+
+    def __call__(self, workflow, input_artifact_filepaths,
+                 parameter_references, output_artifact_filepaths):
+        input_artifact_abs_filepaths = \
+            {k: os.path.abspath(v)
+             for k, v in input_artifact_filepaths.items()}
+        output_artifact_abs_filepaths = \
+            {k: os.path.abspath(v)
+             for k, v in output_artifact_filepaths.items()}
+        job = workflow.to_script(input_artifact_abs_filepaths,
+                                 parameter_references,
+                                 output_artifact_abs_filepaths)
+        with tempfile.NamedTemporaryFile(mode='w', delete=False) as of:
+            of.write(job.code)
+            of.close()
+            cp = subprocess.run([self._python_executable, of.name],
+                                stdout=subprocess.PIPE,
+                                stderr=subprocess.PIPE)
+            os.remove(of.name)
```

Comment:
Ah, fantastic

On Thu, Apr 7, 2016 at 12:51 PM, Greg Caporaso <notifications@github.com>
wrote:

> In qiime/sdk/execution.py
> <https://github.com/biocore/qiime2/pull/40#discussion_r58933751>:
>
> > +        input_artifact_abs_filepaths = \
> > +            {k: os.path.abspath(v)
> > +             for k, v in input_artifact_filepaths.items()}
> > +        output_artifact_abs_filepaths = \
> > +            {k: os.path.abspath(v)
> > +             for k, v in output_artifact_filepaths.items()}
> > +        job = workflow.to_script(input_artifact_abs_filepaths,
> > +                                 parameter_references,
> > +                                 output_artifact_abs_filepaths)
> > +        with tempfile.NamedTemporaryFile(mode='w', delete=False) as of:
> > +            of.write(job.code)
> > +            of.close()
> > +            cp = subprocess.run([self._python_executable, of.name],
> > +                                stdout=subprocess.PIPE,
> > +                                stderr=subprocess.PIPE)
> > +            os.remove(of.name)
>
> @wasade <https://github.com/wasade>, re: your first question
> subprocess.run blocks until the subprocess complete, so I think the
> answer has to be yes. This is now the recommended
> <https://docs.python.org/3/library/subprocess.html> way of using
> subprocess in Python 3.5 and later, so I was going with that. Let me know
> if there is a problem with it though.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/biocore/qiime2/pull/40/files/ea269bd2fc5421e659f4a0c5b9f135de00edec97#r58933751>
>